### ISSUE TITLE
setup.py install for linecache2 syntax error centos 6.7

### ISSUE THREAD

**Issue by centminmod** [letsencrypt/letsencrypt#734]
On CentOS 6.7

running command

    ./venv/bin/pip install -r requirements.txt acme/ . letsencrypt-apache/ letsencrypt-nginx/

I get at the end the following error for

     SyntaxError: ('invalid syntax', ('/root/tools/letsencrypt/venv/lib/python2.6/site-packages/linecache2/tests/inspect_fodder2.py', 102, 25, 'def keyworded(*arg1, arg2=1):\n'))

full excerpt

    Running setup.py install for linecache2
        [pbr] Writing ChangeLog
        [pbr] Generating ChangeLog
        [pbr] ChangeLog complete (0.2s)
        [pbr] Generating AUTHORS
        [pbr] AUTHORS complete (0.4s)
        [pbr] Reusing existing SOURCES.txt
        SyntaxError: ('invalid syntax', ('/root/tools/letsencrypt/venv/lib/python2.6/site-packages/linecache2/tests/inspect_fodder2.py', 102, 25, 'def keyworded(*arg1, arg2=1):\n'))
        
    Successfully installed ConfigArgParse argparse cryptography mock pyrfc3339 ndg-httpsclient pyasn1 PyOpenSSL pytz requests six werkzeug acme configobj parsedatetime psutil python2-pythondialog zope.component zope.interface letsencrypt python-augeas letsencrypt-apache pyparsing letsencrypt-nginx ordereddict unittest2 idna enum34 ipaddress cffi zope.event traceback2 pycparser linecache2
    Cleaning up...

full gist output https://gist.github.com/centminmod/680fcd3cf154e6223a5e

    python2 --version   
    Python 2.6.6

**Comment by centminmod** [comment]
guess it's related to python 2.6.6 ? as i also get this error

    ./venv/bin/letsencrypt --help
    /root/tools/letsencrypt/venv/lib/python2.6/site-packages/cryptography/__init__.py:25: DeprecationWarning: Python 2.6 is no longer supported by the Python core team, please upgrade your Python.
  DeprecationWarning

I have python 2.7.10 installed too

    python2.7 --version            
    Python 2.7.10

so should of used instead of

    virtualenv --no-site-packages -p python2 venv

use

    virtualenv --no-site-packages -p python2.7 venv

?

**Comment by centminmod** [comment]
so too soon got an error doing manual authentication

    ./venv/bin/letsencrypt -d le1.http2ssl.xyz auth

get

    An unexpected error occurred. Please see the logfiles in /var/log/letsencrypt for more details

and log at /var/log/letsencrypt/letsencrypt.log  says

    2015-08-29 05:40:38,204:DEBUG:acme.jose.json_util:Omitted empty fields: challenges=None, combinations=None, status=None, expires=None
    2015-08-29 05:40:38,204:DEBUG:acme.client:Serialized JSON: {"identifier": {"type": "dns", "value": "le1.http2ssl.xyz"}, "resource": "new-authz"}
    2015-08-29 05:40:38,205:DEBUG:acme.jose.json_util:Omitted empty fields: alg=None, x5c=(), crit=(), kid=None, jwk=None, typ=None, jku=None, x5tS256=None, cty=None, x5u=None, x5t=None
    2015-08-29 05:40:38,209:DEBUG:acme.jose.json_util:Omitted empty fields: x5c=(), crit=(), kid=None, typ=None, nonce=None, jku=None, x5tS256=None, cty=None, x5u=None, x5t=None
    2015-08-29 05:40:38,209:DEBUG:root:Sending POST request to https://acme-staging.api.letsencrypt.org/acme/new-authz. args: (), kwargs: {'data': '{"header": {"alg": "RS256", "jwk": {"e": "AQAB", "kty": "RSA", "n": "mw6R4otmOQy9jot4gHd7MvS4TkbCO0U4OswY9PleU6FA_Xu694ReC3zLdrJhjqb4dEs7BQbeR1CGT0W3z69Ngv7hP8T_UKQosL9dlUr2JMechC2WK_q7UmoV5TZjnqE1R6yhbGcLQbEI_QUvsA7hy4fCO7drzu6GrGeZCe1rszmb4xJ4EvcRVT_NEZBXJ6a07PC_jLBn_TNFlsLuXbtOjoQuXX4caahN_rWvHphNqzjbqSR_yztETVAkbYcCkjTYwyfx1vU1TiymVu60uQno317ROTrVj2QE5MsErh_nNRZ8n8HwvHtwH4D2oB7tzX6oaf_3sX0_dh5fFh4nL4BTwQ"}}, "protected": "eyJub25jZSI6ICJIc1ctVzdOWWxscVFzZmRNX1VJR1piUWpwSXY5TWNaQVkyTkI3ZktQRnQ0In0", "payload": "eyJpZGVudGlmaWVyIjogeyJ0eXBlIjogImRucyIsICJ2YWx1ZSI6ICJsZTEuaHR0cDJzc2wueHl6In0sICJyZXNvdXJjZSI6ICJuZXctYXV0aHoifQ", "signature": "Bk72E7-UZqCXRSbLiDH_XsetEjky6FBic9p5Bno5KEAaHkvD_4KwEfihBRBDsXB8HoQifeoQv0c3Mqr-8Jd-gwkNTxG1W0yWMqxCo7qrDSdsdyAmX7N4nkdKBk3X8UWuuvFrjsTlZacw7p9eCg_uDUdGXMbzcOJ8JOt3H0LpY85YU4cyXQ9TZTpL6vP-    M7SOOAvj7EuVTlrVC6km6vcS7DLMD6zG9vcXrUB4xTevcNavC5C8U5NGs9kAjhWlRMmh2qCdUP9LBjvbW8--4s9c-00diaU8ltlkn52jXphmTjIaxAhJiRqt1Jng8TAN9Kp9mB-IU2ysCkT4oA7kS9fENA"}'}
    2015-08-29 05:40:38,210:INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): acme-staging.api.letsencrypt.org
    2015-08-29 05:40:38,898:DEBUG:requests.packages.urllib3.connectionpool:"POST /acme/new-authz HTTP/1.1" 500 76
    2015-08-29 05:40:38,900:DEBUG:root:Received <Response [500]>. Headers: {'content-length': '76', 'expires': 'Sat, 29 Aug 2015 05:40:38 GMT', 'server': 'nginx', 'connection': 'close', 'pragma': 'no-cache', 'cache-control': 'max-age=0, no-cache, no-store', 'date': 'Sat, 29 Aug 2015 05:40:38 GMT', 'access-control-allow-origin': '*', 'content-type': 'application/problem+json', 'replay-nonce':     'fTMMT4wGK5Fq2gCKSpQ4yzlDAcRWwSQyrrRaKzUSywM'}. Content: '{"type":"urn:acme:error:serverInternal","detail":"Error creating new authz"}'
    2015-08-29 05:40:38,901:DEBUG:acme.client:Storing nonce: '}3\x0cO\x8c\x06+\x91j\xda\x00\x8aJ\x948\xcb9C\x01\xc4V\xc1$2\xae\xb4Z+5\x12\xcb\x03'
    2015-08-29 05:40:38,901:DEBUG:acme.client:Received response <Response [500]> (headers: {'content-length': '76', 'expires': 'Sat, 29 Aug 2015 05:40:38 GMT', 'server': 'nginx', 'connection': 'close', 'pragma': 'no-cache', 'cache-control': 'max-age=0, no-cache, no-store', 'date': 'Sat, 29 Aug 2015 05:40:38 GMT', 'access-control-allow-origin': '*', 'content-type': 'application/problem+json', 'replay-nonce':     'fTMMT4wGK5Fq2gCKSpQ4yzlDAcRWwSQyrrRaKzUSywM'}): '{"type":"urn:acme:error:serverInternal","detail":"Error creating new authz"}'
    2015-08-29 05:40:38,903:DEBUG:letsencrypt.cli:Exiting abnormally:
    Traceback (most recent call last):
      File "./venv/bin/letsencrypt", line 9, in <module>
        load_entry_point('letsencrypt==0.1', 'console_scripts', 'letsencrypt')()
      File "/root/tools/letsencrypt/venv/lib/python2.7/site-packages/letsencrypt/cli.py", line 778, in main
        return args.func(args, config, plugins)
      File "/root/tools/letsencrypt/venv/lib/python2.7/site-packages/letsencrypt/cli.py", line 225, in auth
        domains, authenticator, installer, plugins):
      File "/root/tools/letsencrypt/venv/lib/python2.7/site-packages/letsencrypt/client.py", line 240, in obtain_and_enroll_certificate
        certr, chain, key, _ = self.obtain_certificate(domains)
      File "/root/tools/letsencrypt/venv/lib/python2.7/site-packages/letsencrypt/client.py", line 216, in obtain_certificate
        return self._obtain_certificate(domains, csr) + (key, csr)
      File "/root/tools/letsencrypt/venv/lib/python2.7/site-packages/letsencrypt/client.py", line 174, in _obtain_certificate
        authzr = self.auth_handler.get_authorizations(domains)
      File "/root/tools/letsencrypt/venv/lib/python2.7/site-packages/letsencrypt/auth_handler.py", line 73, in get_authorizations
        domain, self.account.regr.new_authzr_uri)
      File "/root/tools/letsencrypt/venv/lib/python2.7/site-packages/acme/client.py", line 202, in request_domain_challenges
        typ=messages.IDENTIFIER_FQDN, value=domain), new_authz_uri)
      File "/root/tools/letsencrypt/venv/lib/python2.7/site-packages/acme/client.py", line 182, in request_challenges
        response = self.net.post(new_authzr_uri, new_authz)
      File "/root/tools/letsencrypt/venv/lib/python2.7/site-packages/acme/client.py", line 596, in post
        return self._check_response(response, content_type=content_type)
      File "/root/tools/letsencrypt/venv/lib/python2.7/site-packages/acme/client.py", line 514, in _check_response
        raise messages.Error.from_json(jobj)
    Error: serverInternal :: The server experienced an internal error :: Error creating new authz

**Comment by kuba** [comment]
You last comment should be filed as a separate issue. Try to `mv /etc/letsencrypt{,.bak}; mv /var/lib/letsencrypt{,.bak}` and retry. Does it work?

It seems that you managed to install it on Python 2.6, so the `SyntaxError` you pasted might be only a warning. `./venv/bin/letsencrypt` keeps working despite deprecation warnings. Is that correct?

**Comment by centminmod** [comment]
@kuba i didn't go that far with Python 2.6, i went straight to trying Python 2.7 for the above

I'll create the last comment as a separate issue :)

**Comment by kuba** [comment]
You pasted the following:

```
./venv/bin/letsencrypt --help
/root/tools/letsencrypt/venv/lib/python2.6/site-packages/cryptography/__init__.py:25: DeprecationWarning: Python 2.6 is no longer supported by the Python core team, please upgrade your Python.
```

which means that you actually managed to install `letsencrypt` in Python 2.6 venv. What happens when you try to run `auth`/`run`/`install`?

**Comment by centminmod** [comment]
With python 2.6.6 can get further for manual authentication as it prompts for

    Make sure your web server displays the following content at
    https://le1.http2ssl.xyz/.well-known/acme-challenge/x7UDVbhNPgMSWU3dagT9jUkvR4sDXiG2zBR0G5z18-Y before continuing:
    
    ǵUYM
    
    Content-Type header MUST be set to application/jose+json.
    
    If you don't have HTTP server configured, you can run the following
    command on the target server (as root):

    mkdir -p /tmp/letsencrypt/public_html/.well-known/acme-challenge
    cd /tmp/letsencrypt/public_html
    echo -n '{"header": {"alg": "RS256", "jwk": {"e": "AQAB", "kty": "RSA", "n": "mw6R4otmOQy9jot4gHd7MvS4TkbCO0U4OswY9PleU6FA_Xu694ReC3zLdrJhjqb4dEs7BQbeR1CGT0W3z69Ngv7hP8T_UKQosL9dlUr2JMechC2WK_q7UmoV5TZjnqE1R6yhbGcLQbEI_QUvsA7hy4fCO7drzu6GrGeZCe1rszmb4xJ4EvcRVT_NEZBXJ6a07PC_jLBn_TNFlsLuXbtOjoQuXX4caahN_rWvHphNqzjbqSR_yztETVAkbYcCkjTYwyfx1vU1TiymVu60uQno317ROTrVj2QE5MsErh_nNRZ8n8HwvHtwH4D2oB7tzX6oaf_3sX0_dh5fFh4nL4BTwQ"}}, "payload": "eyJ0bHMiOiB0cnVlLCAidG9rZW4iOiAieDdVRFZiaE5QZ01TV1UzZGFnVDlqVWt2UjRzRFhpRzJ6QlIwRzV6MTgtWSIsICJ0eXBlIjogInNpbXBsZUh0dHAifQ", "signature": "Ec5Ni5c97NaEp5SnO6nU6dypwNk8eofZe_gAykQMTR01xOXywamFCW0vD67e21NeSMr6ox8IepHGX13Ko2xRCwgPw4tVNUUaNwuDImsZ4lKxnRkUKTp6dFktxeyu2hh4jxHy1vW35eGQxMFs7Qxd4aqGd5Ai1GR6JJtAkLu2ot1R50gsvXNHzWQNUzYoYJqVE-rG8UGFp99loMRsP    -avhoIVw-KZANNX0ycPCukQaECOOXiULXKZa8vRzXUgI5eSvw1ZfZabl7p4emsy6-cgq4trxFXGl3a7SZ7uKW5_7Z-AuVSHJNKMlxKmsUvUkBVIH9EJMMOVcrnNHWdQtIyMjQ"}' > .well-known/acme-challenge/x7UDVbhNPgMSWU3dagT9jUkvR4sDXiG2zBR0G5z18-Y
    # run only once per server:
    openssl req -new -newkey rsa:4096 -subj "/" -days 1 -nodes -x509 -keyout ../key.pem -out ../cert.pem
    $(command -v python2 || command -v python2.7 || command -v python2.6) -c \
    "import BaseHTTPServer, SimpleHTTPServer, ssl; \
    SimpleHTTPServer.SimpleHTTPRequestHandler.extensions_map = {'': 'application/jose+json'}; \
    s = BaseHTTPServer.HTTPServer(('', 443), SimpleHTTPServer.SimpleHTTPRequestHandler); \
    s.socket = ssl.wrap_socket(s.socket, keyfile='../key.pem', certfile='../cert.pem'); \
    s.serve_forever()" 
    Press ENTER to continue

so before i git ENTER I setup my custom Nginx configs challenge file

the nginx vhost for ssl `/usr/local/nginx/conf/conf.d/le1.http2ssl.xyz.conf` I allow dot ./ prefix file

  location ~ /.well-known {
        location ~ /.well-known/acme-challenge/(.*) {
                add_header Content-Type application/jose+json;
        }
  }


    cd /home/nginx/domains/le1.http2ssl.xyz/public
    mkdir -p .well-known/acme-challenge
    echo -n '{"header": {"alg": "RS256", "jwk": {"e": "AQAB", "kty": "RSA", "n": "mw6R4otmOQy9jot4gHd7MvS4TkbCO0U4OswY9PleU6FA_Xu694ReC3zLdrJhjqb4dEs7BQbeR1CGT0W3z69Ngv7hP8T_UKQosL9dlUr2JMechC2WK_q7UmoV5TZjnqE1R6yhbGcLQbEI_QUvsA7hy4fCO7drzu6GrGeZCe1rszmb4xJ4EvcRVT_NEZBXJ6a07PC_jLBn_TNFlsLuXbtOjoQuXX4caahN_rWvHphNqzjbqSR_yztETVAkbYcCkjTYwyfx1vU1TiymVu60uQno317ROTrVj2QE5MsErh_nNRZ8n8HwvHtwH4D2oB7tzX6oaf_3sX0_dh5fFh4nL4BTwQ"}}, "payload": "eyJ0bHMiOiB0cnVlLCAidG9rZW4iOiAieDdVRFZiaE5QZ01TV1UzZGFnVDlqVWt2UjRzRFhpRzJ6QlIwRzV6MTgtWSIsICJ0eXBlIjogInNpbXBsZUh0dHAifQ", "signature": "Ec5Ni5c97NaEp5SnO6nU6dypwNk8eofZe_gAykQMTR01xOXywamFCW0vD67e21NeSMr6ox8IepHGX13Ko2xRCwgPw4tVNUUaNwuDImsZ4lKxnRkUKTp6dFktxeyu2hh4jxHy1vW35eGQxMFs7Qxd4aqGd5Ai1GR6JJtAkLu2ot1R50gsvXNHzWQNUzYoYJqVE-rG8UGFp99loMRsP-avhoIVw-KZANNX0ycPCukQaECOOXiULXKZa8vRzXUgI5eSvw1ZfZabl7p4emsy6-cgq4trxFXGl3a7SZ7uKW5_7Z-AuVSHJNKMlxKmsUvUkBVIH9EJMMOVcrnNHWdQtIyMjQ"}' > .well-known/acme-challenge/x7UDVbhNPgMSWU3dagT9jUkvR4sDXiG2zBR0G5z18-Y

check curl for file

    curl -Ik https://le1.http2ssl.xyz/.well-known/acme-challenge/x7UDVbhNPgMSWU3dagT9jUkvR4sDXiG2zBR0G5z18-Y
    HTTP/1.1 200 OK
    Date: Sat, 29 Aug 2015 07:21:47 GMT
    Content-Type: application/octet-stream
    Content-Length: 911
    Last-Modified: Sat, 29 Aug 2015 07:18:13 GMT
    Connection: keep-alive
    ETag: "55e15cb5-38f"
    Server: nginx centminmod
    Content-Type: application/jose+json
    Accept-Ranges: bytes

then hit ENTER and get

    Press ENTER to continue
    /root/tools/letsencrypt/venv/lib/python2.6/site-packages/requests/packages/urllib3/connectionpool.py:768: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is     strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
      InsecureRequestWarning)
    Incomplete authorizations

FYI, I have a self-signed SSL cert already installed on https://le1.http2ssl.xyz

**Comment by kuba** [comment]
Contents displayed by manual authenticator (`ǵU�YM�` in your example) are not valid. This is fixed by #713.

Until it's fixed you have to make sure your webserver serves whatever is passed to `echo`. In you example this is
```
{"header": {"alg": "RS256", "jwk": {"e": "AQAB", "kty": "RSA", "n": "mw6R4otmOQy9jot4gHd7MvS4TkbCO0U4OswY9PleU6FA_Xu694ReC3zLdrJhjqb4dEs7BQbeR1CGT0W3z69Ngv7hP8T_UKQosL9dlUr2JMechC2WK_q7UmoV5TZjnqE1R6yhbGcLQbEI_QUvsA7hy4fCO7drzu6GrGeZCe1rszmb4xJ4EvcRVT_NEZBXJ6a07PC_jLBn_TNFlsLuXbtOjoQuXX4caahN_rWvHphNqzjbqSR_yztETVAkbYcCkjTYwyfx1vU1TiymVu60uQno317ROTrVj2QE5MsErh_nNRZ8n8HwvHtwH4D2oB7tzX6oaf_3sX0_dh5fFh4nL4BTwQ"}}, "payload": "eyJ0bHMiOiB0cnVlLCAidG9rZW4iOiAieDdVRFZiaE5QZ01TV1UzZGFnVDlqVWt2UjRzRFhpRzJ6QlIwRzV6MTgtWSIsICJ0eXBlIjogInNpbXBsZUh0dHAifQ", "signature": "Ec5Ni5c97NaEp5SnO6nU6dypwNk8eofZe_gAykQMTR01xOXywamFCW0vD67e21NeSMr6ox8IepHGX13Ko2xRCwgPw4tVNUUaNwuDImsZ4lKxnRkUKTp6dFktxeyu2hh4jxHy1vW35eGQxMFs7Qxd4aqGd5Ai1GR6JJtAkLu2ot1R50gsvXNHzWQNUzYoYJqVE-rG8UGFp99loMRsP-avhoIVw-KZANNX0ycPCukQaECOOXiULXKZa8vRzXUgI5eSvw1ZfZabl7p4emsy6-cgq4trxFXGl3a7SZ7uKW5_7Z-AuVSHJNKMlxKmsUvUkBVIH9EJMMOVcrnNHWdQtIyMjQ"}
```

Python 2.6 obviously works here (both installation through `setup.py` and the run itself), so no fixes or updates to the documentation are necessary. I think the issue can be closed now.

**Comment by centminmod** [comment]
yeah it is being served i.e.  https://le1.http2ssl.xyz/.well-known/acme-challenge/x7UDVbhNPgMSWU3dagT9jUkvR4sDXiG2zBR0G5z18-Y

**Comment by kuba** [comment]
Could you please paste full `/var/log/letsencrypt/letsencrypt.log` then?

**Comment by centminmod** [comment]
okay posted all at https://gist.github.com/centminmod/f05ad08b85e3dd71482a

**Comment by centminmod** [comment]
looks like it did create the certs but nothing is said about it after hitting ENTER

    ls -lAhrt /etc/letsencrypt/live/le1.http2ssl.xyz/ 
    total 0
    lrwxrwxrwx 1 root root 43 Aug 29 07:52 privkey.pem -> ../../archive/le1.http2ssl.xyz/privkey1.pem
    lrwxrwxrwx 1 root root 45 Aug 29 07:52 fullchain.pem -> ../../archive/le1.http2ssl.xyz/fullchain1.pem
    lrwxrwxrwx 1 root root 41 Aug 29 07:52 chain.pem -> ../../archive/le1.http2ssl.xyz/chain1.pem
    lrwxrwxrwx 1 root root 40 Aug 29 07:52 cert.pem -> ../../archive/le1.http2ssl.xyz/cert1.pem

**Comment by kuba** [comment]
Urm, I don't see any errors there. Your certificate has been issued. The only thing that might be confusing is the warning about insecure SSL conection, which is expected.

True, there should be some further messages after pressing ENTER...

**Comment by kuba** [comment]
This is what happens for me:
```
Press ENTER to continue
/root/letsencrypt/venv/local/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py:768: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
  InsecureRequestWarning)

IMPORTANT NOTES:
 - If you lose your account credentials, you can recover through
   e-mails sent to me@kuba.us.to.
 - Your account credentials have been saved in your Let's Encrypt
   configuration directory at /etc/letsencrypt. You should make a
   secure backup of this folder now. This configuration directory will
   also contain certificates and private keys obtained by Let's
   Encrypt so making regular backups of this folder is ideal.
 - Automatic renewal and deployment has been enabled for your
   certificate. These settings can be configured in the directories
   under /etc/letsencrypt/configs.
```

Are you sure you don't get the "IMPORTANT NOTES" bit?

**Comment by kuba** [comment]
- https://gist.github.com/centminmod/f05ad08b85e3dd71482a#file-letsencrypt-log-L76
- https://gist.github.com/centminmod/f05ad08b85e3dd71482a#file-letsencrypt-log-L77
- https://gist.github.com/centminmod/f05ad08b85e3dd71482a#file-letsencrypt-log-L133

You should all of those 3 messages.

**Comment by centminmod** [comment]
yes i got those 3 messages at bottom just isn't that clear that SSL certificates have been created.

so maybe a 4th message which should be the first of the four messages as

    - your SSL certificates have been created at /etc/letsencrypt/live/le1.http2ssl.xyz

**Comment by kuba** [comment]
Opened a tracking issue at https://github.com/letsencrypt/letsencrypt/issues/736. Please close this issue, as it's not relevant any more.

**Comment by centminmod** [comment]
thanks @kuba :+1:
